### PR TITLE
Major M-PESA Changes

### PR DESCRIPTION
**Pull request by kamikazechaser**
Safaricom announced the MPES transaction fees will be 100 shillings between registered users (see source below). It was also annonced that any "lipa na mpesa" Transactions below 100 shillings won't attract any fees (see source below). New lipa-na-mpesa section is also created for mmtc-ke which is inline with safaricoms fees conditions. RangeNotFound error is also updated to throw "Consult with mercahnt or service provider", this is because it will not be triggered as the infinity ranges and N/A (-1) combo already prevents a wrong range entry. Instead this error will result for "Dynamic fees" such as lipa na mpesa fees which constantly change depending on merchant or service provider.

Source: http://www.nation.co.ke/news/Safaricom-s-net-profit-jumps-to-Sh23bn-/1056-3440654-fkbub4/index.html

### CODE REVIEW

File: data/mpesa.json [hunk 1]
```diff
@@ -68,6 +67,15 @@
           ]
         },
         {
+          "name": "lipa na mpesa",
+          "ranges": [
+            { "low": "-Infinity", "high": 49, "amount": -1 },
```
Comment (GochoMugo) [inline data/mpesa.json]: Is the `high` correct?

File: data/mpesa.json [hunk 2]
```diff
@@ -68,6 +67,15 @@
...
+            { "low": 10,    "high": 100,   "amount": 0 },
+            { "low": 101,   "high": 70000,  "amount": "Consult merchant" },
```
Comment (GochoMugo) [inline data/mpesa.json]: I would prefer if you set the `amount` to `-2` and add a new property `message` with `Consult merchant`.

Our engine would check if the amount is `-2`. If so, we know we can **not** calculate the amount. Instead, we throw an `AmountNotFoundError` with the provided `message`.

File: engine/math.js [hunk 3]
```diff
@@ -85,7 +85,7 @@ function calculate(name, params) {
     return r !== null && r.low <= amount && amount <= r.high;
   });
   if (!range) {
-    throw new errors.RangeNotFoundError(`range for the amount '${params.amount}' not found`);
+    throw new errors.RangeNotFoundError(`Consult merchant or service provider`);
```
Comment (GochoMugo) [inline engine/math.js]: let's define a new `AmountNotFoundError` that will be thrown if `range.amount` is `-2`. Thus,

```js
if (!range) {
// ... as before ...
}
if (range.amount === -2) {
    throw new errors.AmountNotFoundError(range.message || `cost calculation is not possible`);
}
```

Then propagate the changes...

File: data/mpesa.json (hunk 1 above)
Comment (kamikazechaser) [inline data/mpesa.json]: Oh shit, gotta clean that.

**Comment by kamikazechaser** [comment]
Let me work on this changes

File: engine/errors.js [hunk 4]
```diff
@@ -23,6 +23,7 @@ exports = module.exports = {
   InvalidAmountError: define('InvalidAmountError'),
   NetworkNotFoundError: define('NetworkNotFoundError'),
   RangeNotFoundError: define('RangeNotFoundError'),
+  AmountNotFoundError: define('RangeNotFoundError'),
```
Comment (GochoMugo) [inline engine/errors.js]: Typo! `define("AmountNotFoundError")`

File: engine/math.js [hunk 5]
```diff
@@ -23,6 +23,7 @@ exports = module.exports = {
    *  transaction was not found
    * @throws RangeNotFoundError if the amount was not found in any range
    * @throws InvalidAmountError if the amount entered was invalid
+   * @throws AmountNotFoundError if the amount depends on what the merchant sets
```
Comment (GochoMugo) [inline engine/math.js]:
```js
@throws AmountNotFoundError if the amount can not be calculated
```

The error message will describe why this is the case, for example for lipa-na-mpesa, is `depends on merchant`.

File: engine/math.js [hunk 6]
```diff
@@ -85,8 +86,11 @@ function calculate(name, params) {
     return r !== null && r.low <= amount && amount <= r.high;
   });
   if (!range) {
-    throw new errors.RangeNotFoundError(`Consult merchant or service provider`);
+    throw new errors.RangeNotFoundError(`range for the amount '${params.amount}' not found`);
   }
+  if (range.amount === -2) {
+    throw new errors.AmountNotFoundError(range.message || `Consult merchant`);
```
Comment (GochoMugo) [inline engine/math.js]:
```js
range.message || `Amount can not be calculated`
```

**Pull request merged**